File: skyspark-scala/src/main/scala/com/github/dkanellis/skyspark/scala/api/algorithms/bnl/MedianFinder.scala
package com.github.dkanellis.skyspark.scala.api.algorithms.bnl

import com.github.dkanellis.skyspark.scala.api.algorithms.Point
import org.apache.spark.rdd.RDD

private[bnl] class MedianFinder extends Serializable {

  private[bnl] var numberOfDimensions = 0

  private[bnl] def getMedian(points: RDD[Point]): Point = {
    val numberOfDimensions = points.first().size()

    val medianDimensionValues = new Array[Double](numberOfDimensions)
    for (i <- 0 until numberOfDimensions) {
      medianDimensionValues(i) = getMaxValueOfDimension(points, i) / 2
    }

    Point(medianDimensionValues: _*)
  }

  private def getMaxValueOfDimension(points: RDD[Point], dimensionIndex: Int) = {
    val biggestPointByDimension = points.max()(new Ordering[Point]() {
      override def compare(first: Point, second: Point): Int =
        Ordering[Double].compare(first.dimension(dimensionIndex), second.dimension(dimensionIndex))
    })

    biggestPointByDimension.dimension(dimensionIndex)
  }
}

File: skyspark-java/src/main/java/com/akikanellis/skyspark/api/algorithms/bnl/MedianFinder.java
package com.akikanellis.skyspark.api.algorithms.bnl;

import com.akikanellis.skyspark.api.algorithms.Point;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import scala.Tuple2;

import java.io.Serializable;
import java.util.Collection;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.StreamSupport;

class MedianFinder implements Serializable {

    Point getMedian(JavaRDD<Point> points) {
        JavaPairRDD<Integer, Iterable<Double>> dimensionValuesGrouped = points
                .flatMapToPair(this::extractDimensions)
                .groupByKey();

        JavaPairRDD<Integer, Double> maxesOfEachDimension = dimensionValuesGrouped
                .mapValues(dimensions -> findMax(dimensions) / 2);

        return maxesOfEachDimension
                .coalesce(1)
                .glom()
                .map(this::tuplesToPoint)
                .first();
    }

    private Double findMax(Iterable<Double> dimensions) {
        return StreamSupport.stream(dimensions.spliterator(), true)
                .max(Double::compare)
                .orElseThrow(() -> new IllegalArgumentException("The point was empty"));
    }

    private Iterable<Tuple2<Integer, Double>> extractDimensions(Point point) {
        return IntStream.range(0, point.size())
                .mapToObj(i -> new Tuple2<>(i, point.dimension(i)))
                .collect(Collectors.toList());
    }

    private Point tuplesToPoint(Collection<Tuple2<Integer, Double>> tuples){
        double[] dimensions = new double[tuples.size()];
        tuples.forEach(t -> {
            int index = t._1();
            double value = t._2();

            dimensions[index] = value;
        });

        return new Point(dimensions);
    }
}
